### PR TITLE
New model: BERT-Squad

### PR DESCRIPTION
**Pull request by kundanapillari**
# BERT-Squad

## Use-cases
This model answers questions based on the context of the given input paragraph. 

## Description
BERT (Bidirectional Encoder Representations from Transformers) applies Transformers, a popular attention model, to language modelling. This mechanism has an encoder to read the input text and a decoder that produces a prediction for the task. This model uses the technique of masking out some of the words in the input and then condition each word bidirectionally to predict the masked words. BERT also learns to model relationships between sentences, predicts if the sentences are connected or not.

## Model
 |Model |Download  |Checksum|Download (with sample test data)| ONNX version |Opset version | 
| ------------- |  -------------  | ------------- | ------------- | ------------- | ------------- | 
|BERT-Squad | [393 MB](model/bertsquad8.onnx.tar.gz)  |[MD5](BERT8-md5.txt) |  [788 MB](model/download_sample_8.tar.gz) |  1.3 | 8|
|BERT-Squad| [393 MB](model/bertsquad10.onnx.tar.gz)  |[MD5](BERT10-md5.txt) |  [788 MB](model/download_sample_10.tar.gz) |  1.5 | 10| 

Dependencies
 * [tokenization.py](dependencies/tokenization.py)  
 * [run_onnx_squad.py](dependencies/run_onnx_squad.py)  

## Inference
We used [ONNX Runtime](https://github.com/microsoft/onnxruntime) to perform the inference. 

## Input 
The input is a paragraph and questions relating to that paragraph. The model uses WordPiece tokenisation method to split the input text into list of tokens that are available in the vocabulary (30,522 words).
Then converts these tokens into features
<li>input_ids: list of numerical ids for the tokenised text
<li>input_mask: will be set to 1 for real tokens and 0 for the padding tokens
<li>segment_ids: for our case, this will be set to the list of ones
<li>label_ids: one-hot encoded labels for the text

## Preprocessing
Write an inputs.json file that includes the context paragraph and questions. 
```python
%%writefile inputs.json
{
  "version": "1.4",
  "data": [
    {
      "paragraphs": [
        {
          "context": "In its early years, the new convention center failed to meet attendance and revenue expectations.[12] By 2002, many Silicon Valley businesses were choosing the much larger Moscone Center in San Francisco over the San Jose Convention Center due to the latter's limited space. A ballot measure to finance an expansion via a hotel tax failed to reach the required two-thirds majority to pass. In June 2005, Team San Jose built the South Hall, a $6.77 million, blue and white tent, adding 80,000 square feet (7,400 m2) of exhibit space",
          "qas": [
            {
              "question": "where is the businesses choosing to go?",
              "id": "1"
            },
            {
              "question": "how may votes did the ballot measure need?",
              "id": "2"
            },
            {
              "question": "By what year many Silicon Valley businesses were choosing the Moscone Center?",
              "id": "3"
            }
          ]
        }
      ],
      "title": "Conference Center"
    }
  ]
}
```
Get parameters and convert input examples into features
```python
# preprocess input 
predict_file = 'inputs.json'

# Use read_squad_examples method from run_onnx_squad to read the input file
eval_examples = read_squad_examples(input_file=predict_file)

max_seq_length = 256
doc_stride = 128
max_query_length = 64
batch_size = 1
n_best_size = 20
max_answer_length = 30

vocab_file = os.path.join('uncased_L-12_H-768_A-12', 'vocab.txt')
tokenizer = tokenization.FullTokenizer(vocab_file=vocab_file, do_lower_case=True)

# Use convert_examples_to_features method from run_onnx_squad to get parameters from the input 
input_ids, input_mask, segment_ids, extra_data = convert_examples_to_features(eval_examples, tokenizer,
                                                                              max_seq_length, doc_stride, max_query_length)
```

## Output
For each question about the context paragraph, the model predicts a start and an end token from the paragraph that most likely answers the questions. 

## Postprocessing
Write the predictions (answers to the questions) in a file. 
```python
# postprocess results
output_dir = 'predictions'
os.makedirs(output_dir, exist_ok=True)
output_prediction_file = os.path.join(output_dir, "predictions.json")
output_nbest_file = os.path.join(output_dir, "nbest_predictions.json")
write_predictions(eval_examples, extra_data, all_results,
                  n_best_size, max_answer_length,
                  True, output_prediction_file, output_nbest_file)
```

## Dataset (Train and Validation)
The model is trained with [SQuAD v1.1](https://rajpurkar.github.io/SQuAD-explorer/explore/1.1/dev/) dataset that contains 100,000+ question-answer pairs on 500+ articles. 

## Validation accuracy
Metric is Exact Matching (EM) of 80.7, computed over SQuAD v1.1 dev data, for this onnx model. 

## Training
Fine-tuned the model using SQuAD-1.1 dataset. Look at [BertTutorial.ipynb](https://github.com/onnx/tensorflow-onnx/blob/master/tutorials/BertTutorial.ipynb) for more information for converting the model from tensorflow to onnx and for fine-tuning.

## References
* **BERT** Model from the paper [BERT: Pre-training of Deep Bidirectional Transformers for Language Understanding](https://arxiv.org/abs/1810.04805)

* [BERTtutorial](https://github.com/onnx/tensorflow-onnx/blob/master/tutorials/BertTutorial.ipynb)

## Contributors
[Kundana Pillari](https://github.com/kundanapillari)

## License
Apache 2.0
<hr>

### CODE REVIEW

File: text/machine_comprehension/bert-squad/README.md [hunk 1]
```diff
@@ -0,0 +1,111 @@
+# BERT-Squad
+
+## Use cases
+This model uses masked language method to enable training of bidirectional models that can quickly and efficiently run through language processing tasks. The model also adds a next sentence prediction task to improve sentence-level understanding which helps to answer questions based on the given paragraph input.
```
Comment (prasanthpul) [inline text/machine_comprehension/bert-squad/README.md]: This sounds like how the model is trained. Use case should describe why someone would use this model, what scenario is addresses (for example, for resnet, it's classifying images into one of 1000 categories.)

File: text/machine_comprehension/bert-squad/README.md [hunk 2]
```diff
@@ -0,0 +1,111 @@
+# BERT-Squad
+
+## Use cases
+This model uses masked language method to enable training of bidirectional models that can quickly and efficiently run through language processing tasks. The model also adds a next sentence prediction task to improve sentence-level understanding which helps to answer questions based on the given paragraph input. 
+
+## Description
+BERT (Bidirectional Encoder Representations from Transformers) applies Transformers, a popular attention model, to language modelling. This mechanism has an encoder to read the input text and a decoder that produces a prediction for the task. This model uses the technique of masking out some of the words in the input and then condition each word bidirectionally to predict the masked words. BERT also learns to model relationships between sentences, predicts if the sentences are connected or not.
```
Comment (prasanthpul) [inline text/machine_comprehension/bert-squad/README.md]: Need to list the accuracy of this model on the Squad dataset.

File: text/machine_comprehension/bert-squad/README.md [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+ |Model        |Download  |Checksum|Download (with sample test data)| ONNX version |Opset version|Dependencies| 
+| ------------- | ------------- | ------------- | ------------- | ------------- | ------------- | ------------- |
+|BERT-Squad| [393,930 KB](model/bert.onnx)  |dc59794145b4a37995a2667370dc3d6f |  [78 KB](model/saved_model.pb.gz) |  1.5.2  | 8| [tokenization.py](dependencies/tokenization.py) [run_onnx_squad.py](dependencies/run_onnx_squad.py)  | 
+
+## Inference
+We used onnxruntime to preform the inference. 
+
+### Input 
+Context paragraph and questions about the paragraph that are written in an inputs.json file.
```
Comment (prasanthpul) [inline text/machine_comprehension/bert-squad/README.md]: Input should describe the input the actual model takes which are probably tensors that represent the the context paragraph and questions. The inputs.json is just how the preprocessing code we are using works. If someone used this model they would most likely not use an inputs.json, they would feed in the data directly.

File: text/machine_comprehension/bert-squad/README.md [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+vocab_file = os.path.join('uncased_L-12_H-768_A-12', 'vocab.txt')
+tokenizer = tokenization.FullTokenizer(vocab_file=vocab_file, do_lower_case=True)
+
+# Use convert_examples_to_features method from run_onnx_squad to get parameters from the input 
+input_ids, input_mask, segment_ids, extra_data = convert_examples_to_features(eval_examples, tokenizer,
+                                                                              max_seq_length, doc_stride, max_query_length)
+```
+
+### Output
+The model produces the predictions/answers for the questions asked based on context from the given input paragraph.
```
Comment (prasanthpul) [inline text/machine_comprehension/bert-squad/README.md]: Output is also some tensors. How are they transformed into the predictions/answers? There is probably some vocabulary file involved

File: text/machine_comprehension/bert-squad/README.md [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+write_predictions(eval_examples, extra_data, all_results,
+                  n_best_size, max_answer_length,
+                  True, output_prediction_file, output_nbest_file)
+```
+
+## Dataset
+The model is trained with [SQuAD v1.1](https://rajpurkar.github.io/SQuAD-explorer/explore/1.1/dev/) dataset that contains 100,000+ question-answer pairs on 500+ articles. 
+
+## Validation accuracy
+The accuracies obtained by the model on the validation set are mentioned above. The accuracies have been calculated on center cropped images with a maximum deviation of 1% (top-1 accuracy) from the paper.
```
Comment (prasanthpul) [inline text/machine_comprehension/bert-squad/README.md]: How do center cropped images relate to BERT?

File: text/machine_comprehension/bert-squad/README.md [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+
+## Training
+Fine-tuned the model using SQuAD-1.1 dataset. Look at [BertTutorial.ipynb](https://github.com/onnx/tensorflow-onnx/blob/master/tutorials/BertTutorial.ipynb) for more information for converting the model from tensorflow to onnx and for fine-tuning
+
+## Validation
+
```
Comment (prasanthpul) [inline text/machine_comprehension/bert-squad/README.md]: Missing content?

File: text/machine_comprehension/bert-squad/README.md [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+## References
+* **BERT** Model from the paper [BERT: Pre-training of Deep Bidirectional Transformers for Language Understanding](https://arxiv.org/abs/1810.04805)
+
+* [BERTtutorial](https://github.com/onnx/tensorflow-onnx/blob/master/tutorials/BertTutorial.ipynb)
+## Contributors
+[Kundana Pillari](https://github.com/kundanapillari)
+
+## License
+Apache 2.0
```
Comment (prasanthpul) [inline text/machine_comprehension/bert-squad/README.md]: What is this license based on?

File: text/machine_comprehension/bert-squad/README.md (hunk 2 above)
Comment (kundanapillari) [reply text/machine_comprehension/bert-squad/README.md]: I added the accuracy in the validation accuracy section.

File: text/machine_comprehension/bert-squad/README.md [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+
+ |Model        |Download  |Checksum|Download (with sample test data)| ONNX version |Opset version|Dependencies| 
+| ------------- | ------------- | ------------- | ------------- | ------------- | ------------- | ------------- |
+|BERT-Squad| [393,930 KB](model/bert.onnx)  |[MD5](BERT-md5.txt) |  [932 KB](model/saved_model.pb) |  1.5.2  | 8| [tokenization.py](dependencies/tokenization.py) [run_onnx_squad.py](dependencies/run_onnx_squad.py)  | 
+
+## Inference
+We used onnxruntime to preform the inference. 
+
+### Input 
+Uses WordPiece tokenisation method to split the input text into list of tokens that are available in the vocabulary (30,522 words).
```
Comment (prasanthpul) [inline text/machine_comprehension/bert-squad/README.md]: thanks for adding this info about the inputs to the model. you should still keep a description of the input (i.e. paragraph and questions about it)

File: text/machine_comprehension/bert-squad/README.md [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+Then converts these tokens into features
+<li>input_ids: list of numerical ids for the tokenised text
+<li>input_mask: will be set to 1 for real tokens and 0 for the padding tokens
+<li>segment_ids: for our case, this will be set to the list of ones
+<li>label_ids: one-hot encoded labels for the text
+
+### Preprocessing
+Get parameters and convert input examples into features
+```python
```
Comment (prasanthpul) [inline text/machine_comprehension/bert-squad/README.md]: should include the previous example entry for inputs.json

File: text/machine_comprehension/bert-squad/README.md [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+vocab_file = os.path.join('uncased_L-12_H-768_A-12', 'vocab.txt')
+tokenizer = tokenization.FullTokenizer(vocab_file=vocab_file, do_lower_case=True)
+
+# Use convert_examples_to_features method from run_onnx_squad to get parameters from the input 
+input_ids, input_mask, segment_ids, extra_data = convert_examples_to_features(eval_examples, tokenizer,
+                                                                              max_seq_length, doc_stride, max_query_length)
+```
+
+### Output
+Given a question and a context paragraph, the model predicts a start and an end token from the paragraph that most likely answers the questions. Computes a start vector and a final transformer output for each token.
```
Comment (prasanthpul) [inline text/machine_comprehension/bert-squad/README.md]: I would rephrase the first sentence as "For each question about the context paragraph, the model predicts a start and an end token from the paragraph that most likely answers the questions."
I wasnt sure what the second sentence means?

File: text/machine_comprehension/bert-squad/README.md [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+# Use convert_examples_to_features method from run_onnx_squad to get parameters from the input 
+input_ids, input_mask, segment_ids, extra_data = convert_examples_to_features(eval_examples, tokenizer,
+                                                                              max_seq_length, doc_stride, max_query_length)
+```
+
+### Output
+Given a question and a context paragraph, the model predicts a start and an end token from the paragraph that most likely answers the questions. Computes a start vector and a final transformer output for each token. 
+
+### Postprocessing
+Write the predictions (answers to the questions) in a file.
```
Comment (prasanthpul) [inline text/machine_comprehension/bert-squad/README.md]: Postprocessing should be about how to turn the output of the model into something you can use. This should be independent of file or not. In fact it would be much better if we didn;t have to write out to a file.

File: text/machine_comprehension/bert-squad/README.md [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+write_predictions(eval_examples, extra_data, all_results,
+                  n_best_size, max_answer_length,
+                  True, output_prediction_file, output_nbest_file)
+```
+
+## Dataset (Train and Validation)
+The model is trained with [SQuAD v1.1](https://rajpurkar.github.io/SQuAD-explorer/explore/1.1/dev/) dataset that contains 100,000+ question-answer pairs on 500+ articles. 
+
+## Validation accuracy
+Metric is Exact Matching (EM) of 84.3, computed over SQuAD v1.1 dev data.
```
Comment (prasanthpul) [inline text/machine_comprehension/bert-squad/README.md]: Where did this number come from?

File: text/machine_comprehension/bert-squad/README.md [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+
+## References
+* **BERT** Model from the paper [BERT: Pre-training of Deep Bidirectional Transformers for Language Understanding](https://arxiv.org/abs/1810.04805)
+
+* [BERTtutorial](https://github.com/onnx/tensorflow-onnx/blob/master/tutorials/BertTutorial.ipynb)
+## Contributors
+[Kundana Pillari](https://github.com/kundanapillari)
+
+## License
+Add License information
```
Comment (prasanthpul) [inline text/machine_comprehension/bert-squad/README.md]: what license should be used?

Comment (BowenBao) [reply text/machine_comprehension/bert-squad/README.md]: FYI the original tensorflow models are under Apache 2.0

https://github.com/google-research/bert#what-license-is-this-library-released-under

File: text/machine_comprehension/bert-squad/README.md (hunk 12 above)
Comment (kundanapillari) [reply text/machine_comprehension/bert-squad/README.md]: It is from the google-research github for the BERT uncased model because that is the model I trained with squad-1.1 dataset.

Comment (prasanthpul) [reply text/machine_comprehension/bert-squad/README.md]: is it possible to verify this against the ONNX model?

File: text/machine_comprehension/bert-squad/README.md [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+This model answers questions based on the context of the given input paragraph.   
+
+## Description
+BERT (Bidirectional Encoder Representations from Transformers) applies Transformers, a popular attention model, to language modelling. This mechanism has an encoder to read the input text and a decoder that produces a prediction for the task. This model uses the technique of masking out some of the words in the input and then condition each word bidirectionally to predict the masked words. BERT also learns to model relationships between sentences, predicts if the sentences are connected or not. 
+
+## Model
+
+ |Model        |Download  |Checksum|Download (with sample test data)| ONNX version |Opset version|Dependencies| 
+| ------------- | ------------- | ------------- | ------------- | ------------- | ------------- | ------------- |
+|BERT-Squad| [393,930 KB](model/bert.onnx)  |[MD5](BERT-md5.txt) |  [932 KB](model/saved_model.pb) |  1.5.2  | 8| [tokenization.py](dependencies/tokenization.py) [run_onnx_squad.py](dependencies/run_onnx_squad.py)  |
```
Comment (prasanthpul) [inline text/machine_comprehension/bert-squad/README.md]: The "Download (with sample test data)" link should be to a tar.gz file that has the model + the input and output data files.

File: text/machine_comprehension/bert-squad/README.md [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+BERT (Bidirectional Encoder Representations from Transformers) applies Transformers, a popular attention model, to language modelling. This mechanism has an encoder to read the input text and a decoder that produces a prediction for the task. This model uses the technique of masking out some of the words in the input and then condition each word bidirectionally to predict the masked words. BERT also learns to model relationships between sentences, predicts if the sentences are connected or not. 
+
+## Model
+
+ |Model        |Download  |Checksum|Download (with sample test data)| ONNX version |Opset version|Dependencies| 
+| ------------- | ------------- | ------------- | ------------- | ------------- | ------------- | ------------- |
+|BERT-Squad| [393,930 KB](model/bert.onnx)  |[MD5](BERT-md5.txt) |  [932 KB](model/saved_model.pb) |  1.5.2  | 8| [tokenization.py](dependencies/tokenization.py) [run_onnx_squad.py](dependencies/run_onnx_squad.py)  | 
+
+## Inference
+We used onnxruntime to preform the inference.
```
Comment (prasanthpul) [inline text/machine_comprehension/bert-squad/README.md]: preform -> perform

File: text/machine_comprehension/bert-squad/README.md (hunk 12 above)
Comment (kundanapillari) [reply text/machine_comprehension/bert-squad/README.md]: I verified it against the ONNX model and updated the metric

File: text/machine_comprehension/bert-squad/README.md [hunk 16]
```diff
@@ -0,0 +1,115 @@
...
+This model answers questions based on the context of the given input paragraph.   
+
+## Description
+BERT (Bidirectional Encoder Representations from Transformers) applies Transformers, a popular attention model, to language modelling. This mechanism has an encoder to read the input text and a decoder that produces a prediction for the task. This model uses the technique of masking out some of the words in the input and then condition each word bidirectionally to predict the masked words. BERT also learns to model relationships between sentences, predicts if the sentences are connected or not. 
+
+## Model
+
+ |Model        |Download  |Checksum|Download (with sample test data)| ONNX version |Opset version|Dependencies| 
+| ------------- | ------------- | ------------- | ------------- | ------------- | ------------- | ------------- |
+|BERT-Squad| [393 MB](model/bert.onnx.tar.gz)  |[MD5](BERT-md5.txt) |  [394 KB](model/download_sample.tar.gz) |  1.5.2  | 8| [tokenization.py](dependencies/tokenization.py) [run_onnx_squad.py](dependencies/run_onnx_squad.py)  |
```
Comment (prasanthpul) [inline text/machine_comprehension/bert-squad/README.md]: If it's opset 8, then it should be ONNX 1.3

File: text/machine_comprehension/bert-squad/README.md [hunk 17]
```diff
@@ -0,0 +1,115 @@
...
+BERT (Bidirectional Encoder Representations from Transformers) applies Transformers, a popular attention model, to language modelling. This mechanism has an encoder to read the input text and a decoder that produces a prediction for the task. This model uses the technique of masking out some of the words in the input and then condition each word bidirectionally to predict the masked words. BERT also learns to model relationships between sentences, predicts if the sentences are connected or not. 
+
+## Model
+
+ |Model        |Download  |Checksum|Download (with sample test data)| ONNX version |Opset version|Dependencies| 
+| ------------- | ------------- | ------------- | ------------- | ------------- | ------------- | ------------- |
+|BERT-Squad| [393 MB](model/bert.onnx.tar.gz)  |[MD5](BERT-md5.txt) |  [394 KB](model/download_sample.tar.gz) |  1.5.2  | 8| [tokenization.py](dependencies/tokenization.py) [run_onnx_squad.py](dependencies/run_onnx_squad.py)  | 
+
+## Inference
+We used onnxruntime to perform the inference.
```
Comment (prasanthpul) [inline text/machine_comprehension/bert-squad/README.md]: "onnxruntime" -> [ONNX Runtime](https://github.com/microsoft/onnxruntime)

**Comment by kundanapillari** [comment]
> [CLA assistant check](https://cla-assistant.io/onnx/models?pullRequest=190) 
> Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/onnx/models?pullRequest=190) before we can accept your contribution.
> **1** out of **2** committers have signed the CLA.
> 
> ✅ kundanapillari
> ❌ Kundana Pillari
> 
> **Kundana Pillari** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).
> You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/onnx/models?pullRequest=190) it.

File: text/machine_comprehension/bert-squad/README.md [hunk 18]
```diff
@@ -0,0 +1,120 @@
...
+This model answers questions based on the context of the given input paragraph.   
+
+## Description
+BERT (Bidirectional Encoder Representations from Transformers) applies Transformers, a popular attention model, to language modelling. This mechanism has an encoder to read the input text and a decoder that produces a prediction for the task. This model uses the technique of masking out some of the words in the input and then condition each word bidirectionally to predict the masked words. BERT also learns to model relationships between sentences, predicts if the sentences are connected or not. 
+
+## Model
+
+ |Model        |Download  |Checksum|Download (with sample test data)| ONNX version |Opset version|
+| ------------- | ------------- | ------------- | ------------- | ------------- | ------------- | 
+|BERT-Squad| [393 MB](model/bert8.onnx.tar.gz)  |[MD5](BERT-md5-8.txt) |  [788 MB](model/download_sample_8.tar.gz) |  1.3 | 8|
```
Comment (WilliamTambellini) [inline text/machine_comprehension/bert-squad/README.md]: Could you please rename the file "bertsquad8.onnx.tar.gz" (etc) ?

File: text/machine_comprehension/bert-squad/README.md [hunk 19]
```diff
@@ -0,0 +1,120 @@
...
+|BERT-Squad| [393 MB](model/bert10.onnx.tar.gz)  |[MD5](BERT-md5-10.txt) |  [788 MB](model/download_sample_10.tar.gz) |  1.5 | 10|
```
Comment (WilliamTambellini) [inline text/machine_comprehension/bert-squad/README.md]: bertsquad10.onnx ?

**Comment by GeorgeS2019** [comment]
@kundanapillari  ONNX needs more BERT related pre-trained models (e.g. other languages)

**Comment by GeorgeS2019** [comment]
@kundanapillari   Can someone show how to get [this ](https://github.com/onnx/models/issues/130#issuecomment-523149760)work? PyTorch to ONNX 

For PyTorch, example of exporting ONNX model and inference using ONNX Runtime:
[https://github.com/tianleiwu/pytorch-pretrained-BERT/blob/master/examples/run_classifier.py](https://github.com/tianleiwu/pytorch-pretrained-BERT/blob/master/examples/run_classifier.py)
